### DISCUSSION TITLE
Pyright behavior with generic descriptors and protocols

### DISCUSSION THREAD

**Discussion by CaselIT** [microsoft/pyright#5915] [Q&A]
Hi,

I've noticed a difference between how mypy and pyright handles descriptors when using protocols.
I'm creating a discussion since I'm not sure if mypy or pyright has the correct behavior.

I noticed this in sqlalchemy, in the example posted in this issue https://github.com/sqlalchemy/sqlalchemy/issues/10288.

Let's start with a setup of a descriptor that's equivalent to the one used by sqlalchemy
```py
from typing import Any, Generic, TypeVar, overload

_T_co = TypeVar("_T_co", covariant=True)

class InstrumentedAttribute(Generic[_T_co]):
    '''The real version has a lot of methods'''
    def something(self) -> _T_co:
        ...

class Mapped(Generic[_T_co]):
    @overload
    def __get__(
        self, instance: None, owner: Any
    ) -> InstrumentedAttribute[_T_co]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T_co:
        ...

    def __get__(
        self, instance: object | None, owner: Any
    ) -> InstrumentedAttribute[_T_co] | _T_co:
        ...
```
Let's also define a couple of classes and a protocol
```py
from typing import Protocol

class ParentProtocol(Protocol):
    name: Mapped[str]

class Parent:
    name: Mapped[str]

class Item:
    id: Mapped[int]
    parent: Mapped[Parent]
```

1. First question
    ```py
    def check_parent_protocol(p: ParentProtocol) -> str:
        return p.name


    def check_parent(p: Parent) -> str:
        return p.name


    check_parent(Parent) # error
    check_parent(Parent())  # ok
    check_parent_protocol(Parent) # ???
    check_parent_protocol(Parent())  # ok
    ```

    pyright reports the following
    ```
    test1.py:57:14 - error: Argument of type "type[Parent]" cannot be assigned to parameter "p" of type "Parent" in function "check_parent"
       "type[type]" is incompatible with "type[Parent]" (reportGeneralTypeIssues)
    ```

    while mypy reports
    ```
    test1.py:57: error: Argument 1 to "check_parent" has incompatible type "type[Parent]"; expected "Parent"  [arg-type]
    test1.py:59: error: Argument 1 to "check_parent_protocol" has incompatible type "type[Parent]"; expected "ParentProtocol"  [arg-type]
    test1.py:59: note: Only class variables allowed for class object access on protocols, name is an instance variable of "Parent"
    ```

    It seems that pyright has a different behavior when using a protocol versus a class. I'm not sure what the correct behaviour is in this case, but different behaviour seems strange.

2. Second question
    ```py
    class PropProtocol(Protocol):
        @property
        def parent(self) -> Mapped[ParentProtocol]:
            ...


    def get_parent_name(child: PropProtocol) -> str:
        reveal_type(child.parent)
        return child.parent.name


    get_parent_name(Item)
    get_parent_name(Item())
    ```

    pyright reports the following
    ```
    test1.py:70:17 - information: Type of "child.parent" is "Mapped[ParentProtocol]"
    test1.py:71:25 - error: Cannot access member "name" for type "Mapped[ParentProtocol]"
      Member "name" is unknown (reportGeneralTypeIssues)
    ```

    while mypy does this
    ```
    test1.py:70: note: Revealed type is "test1.ParentProtocol"
    test1.py:74: error: Argument 1 to "get_parent_name" has incompatible type "type[Item]"; expected "PropProtocol"  [arg-type]
    test1.py:74: note: Only class variables allowed for class object access on protocols, parent is an instance variable of "Item"
    ```

    Here the difference are two:
    - as before pyright allows the type in `get_parent_name(Item)` since the function is typed with a protocol. Same consideration as above
    - pyright interprets the property as returning a descriptor, so it considers `child.parent` a descriptor instance, while mypy uses the gets defined in the descriptor.
    Not sure what's the correct behaviour here, but since pyright accepts `Item()` as a `PropProtocol` it seems that the correct behaviour should be the mypy one

3. Third question
    ```py
    class AttrProtocol(Protocol):
        parent: Mapped[ParentProtocol]


    def get_parent_name_again(child: AttrProtocol) -> str:
        return child.parent.name


    get_parent_name_again(Item)
    get_parent_name_again(Item())
    ```

    pyright reports the following
    ```
    test1.py:86:23 - error: Argument of type "type[Item]" cannot be assigned to parameter "child" of type "AttrProtocol" in function "get_parent_name_again"
      "parent" is invariant because it is mutable
      "parent" is an incompatible type
        "Mapped[Parent]" is incompatible with "Mapped[ParentProtocol]"
          Type parameter "_T_co@Mapped" is covariant, but "Parent" is not a subtype of "ParentProtocol"
            "Parent" is incompatible with protocol "ParentProtocol" (reportGeneralTypeIssues)
    test1.py:87:23 - error: Argument of type "Item" cannot be assigned to parameter "child" of type "AttrProtocol" in function "get_parent_name_again"
      "Item" is incompatible with protocol "AttrProtocol"
        "parent" is invariant because it is mutable
        "parent" is an incompatible type
          "Mapped[Parent]" is incompatible with "Mapped[ParentProtocol]"
            Type parameter "_T_co@Mapped" is covariant, but "Parent" is not a subtype of "ParentProtocol"
              "Parent" is incompatible with protocol "ParentProtocol" (reportGeneralTypeIssues)
    ```

    mypy instead reports
    ```
    test1.py:86: error: Argument 1 to "get_parent_name_again" has incompatible type "type[Item]"; expected "AttrProtocol"  [arg-type]
    test1.py:86: note: Following member(s) of "Item" have conflicts:
    test1.py:86: note:     parent: expected "Mapped[ParentProtocol]", got "Mapped[Parent]"
    test1.py:86: note: Only class variables allowed for class object access on protocols, parent is an instance variable of "Item"
    test1.py:87: error: Argument 1 to "get_parent_name_again" has incompatible type "Item"; expected "AttrProtocol"  [arg-type]
    test1.py:87: note: Following member(s) of "Item" have conflicts:
    test1.py:87: note:     parent: expected "Mapped[ParentProtocol]", got "Mapped[Parent]"
    ```

    Here there seems to a bug in pyright since it reports `Type parameter "_T_co@Mapped" is covariant, but "Parent" is not a subtype of "ParentProtocol"` but previously it was fine in `check_parent`

    Otherwise I'm not sure if AttrProtocol should be preferred over PropProtocol. It seems to me that it should be the least ambiguous way of defining a protocol, but since nighter mypy nor pyright likes it suggests that it's not.

Sorry for the long post, but I kept finding differences in this case. Thanks for reading

<details><summary>Complete file to reproduce</summary>
<p>

```py
# mypy: disable-error-code="empty-body"
from typing import Any, Generic, TypeVar, overload

_T_co = TypeVar("_T_co", covariant=True)


class InstrumentedAttribute(Generic[_T_co]):
    def something(self) -> _T_co:
        ...

    # many other methods


class Mapped(Generic[_T_co]):
    @overload
    def __get__(
        self, instance: None, owner: Any
    ) -> InstrumentedAttribute[_T_co]:
        ...

    @overload
    def __get__(self, instance: object, owner: Any) -> _T_co:
        ...

    def __get__(
        self, instance: object | None, owner: Any
    ) -> InstrumentedAttribute[_T_co] | _T_co:
        ...


from typing import Protocol


class ParentProtocol(Protocol):
    name: Mapped[str]


class Parent:
    name: Mapped[str]


class Item:
    id: Mapped[int]
    parent: Mapped[Parent]


def check_parent_protocol(p: ParentProtocol) -> str:
    return p.name


def check_parent(p: Parent) -> str:
    return p.name


check_parent(Parent)
check_parent(Parent())  # ok
check_parent_protocol(Parent)
check_parent_protocol(Parent())  # ok


class PropProtocol(Protocol):
    @property
    def parent(self) -> Mapped[ParentProtocol]:
        ...


def get_parent_name(child: PropProtocol) -> str:
    reveal_type(child.parent)
    return child.parent.name


get_parent_name(Item)
get_parent_name(Item())


class AttrProtocol(Protocol):
    parent: Mapped[ParentProtocol]


def get_parent_name_again(child: AttrProtocol) -> str:
    return child.parent.name


get_parent_name_again(Item)
get_parent_name_again(Item())

```

</p>
</details>

**Comment by erictraut** [comment]
**Question 1**
This question involves the difference between nominal and structural types. A nominal type uses class hierarchies to determine type compatibility. The expression `Parent` has the type `type[Parent]` and is not a subtype of `Parent`, so this is not type compatible. Structural types (protocols) work differently. They define an interface, and a type is deemed compatible with the structural type if it complies with that interface. The `ParentProtocol` class defines an interface that includes a single attribute named `name` with a type of `Mapped[str]`. Values that comply with that interface are compatible. Instances of `Parent` surely satisfy this interface, but so does the class `Parent` itself. That's because it has a class variable named `name` with a type that is compatible with `Mapped[str]`.

Mypy and pyright differ slightly in how they interpret class variables vs instance variables. Mypy, in particular, is a bit sloppy and inconsistent in how it handles these. There are multiple open issues in the mypy issue tracker on this topic. You can read more about this [here](https://microsoft.github.io/pyright/#/mypy-comparison?id=class-and-instance-variable-enforcement). I suspect that the reason mypy rejects `check_parent_protocol(Parent)` is because it treats `ParentProtocol.name` as a [pure instance variable](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=pure-instance-variables) as opposed to a [regular class variable](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=regular-class-variables).

**Question 2**
When a descriptor (or a property, which is implemented as a descriptor) returns another descriptor, the runtime doesn't invoke the descriptor protocol a second time. Doing so could lead to infinite recursion. The runtime therefore invokes a descriptor only once. If that descriptor returns another descriptor, it doesn't invoke the second descriptor's `__get__` method. Mypy appears to not model this correctly. Here's a simplified example that demonstrates this.

```python
from typing import Any, reveal_type

class IntDescriptor:
    def __get__(self, instance: object | None, owner: Any) -> int:
        return 0

class Foo:
    @property
    def parent(self) -> IntDescriptor:
        return IntDescriptor()

f = Foo()
reveal_type(f.parent)  # pyright reveals "IntDescriptor", mypy reveals "int"
print(f.parent) # Runtime prints "IntDescriptor, which is consistent with pyright
```

**Question 3**
Mypy and pyright agree here. These are both type errors.

Pyright's nested error message is a bit misleading here. While `Mapped[Parent]` is a subtype of `Mapped[ParentProtocol]`, the `parent` field is invariant because it is mutable. In light of this invariance, `Mapped[Parent]` and `Mapped[ParentProtocol]` are not compatible types.

I'll look at improving the error message to eliminate the confusion.

**Reply by erictraut** [reply]
I've improved the error message related to question 3. This will be included in the next release of pyright.

**Reply by CaselIT** [reply]
Thanks for the reply. From your answer it seems that the correct way to type this would be one of these
```py
class Prop2Protocol(Protocol):
    @property
    def parent(self) -> ParentProtocol:
        ...


def get_parent_name_prop2(child: Prop2Protocol) -> str:
    reveal_type(child.parent)
    return child.parent.name


get_parent_name_prop2(Item)
get_parent_name_prop2(Item())


class Attr2Protocol(Protocol):
    parent: ParentProtocol


def get_parent_name_attr2(child: Attr2Protocol) -> str:
    return child.parent.name


get_parent_name_attr2(Item)
get_parent_name_attr2(Item())
```

but neither is accepted. pyright
```
test1.py:110:23 - error: Argument of type "type[Item]" cannot be assigned to parameter "child" of type "Attr2Protocol" in function "get_parent_name_attr2"
    "parent" is invariant because it is mutable
    "parent" is an incompatible type
      "Mapped[Parent]" is incompatible with protocol "ParentProtocol" (reportGeneralTypeIssues)
test1.py:111:23 - error: Argument of type "Item" cannot be assigned to parameter "child" of type "Attr2Protocol" in function "get_parent_name_attr2"
    "Item" is incompatible with protocol "Attr2Protocol"
      "parent" is invariant because it is mutable
      "parent" is an incompatible type
        "Mapped[Parent]" is incompatible with protocol "ParentProtocol" (reportGeneralTypeIssues)
```
mypy
```
test1.py:98: error: Argument 1 to "get_parent_name_prop2" has incompatible type "type[Item]"; expected "Prop2Protocol"  [arg-type]
test1.py:98: note: Following member(s) of "Item" have conflicts:
test1.py:98: note:     parent: expected "ParentProtocol", got "Mapped[Parent]"
test1.py:98: note: Only class variables allowed for class object access on protocols, parent is an instance variable of "Item"
test1.py:99: error: Argument 1 to "get_parent_name_prop2" has incompatible type "Item"; expected "Prop2Protocol"  [arg-type]
test1.py:99: note: Following member(s) of "Item" have conflicts:
test1.py:99: note:     parent: expected "ParentProtocol", got "Mapped[Parent]"
test1.py:110: error: Argument 1 to "get_parent_name_attr2" has incompatible type "type[Item]"; expected "Attr2Protocol"  [arg-type]
test1.py:110: note: Following member(s) of "Item" have conflicts:
test1.py:110: note:     parent: expected "ParentProtocol", got "Mapped[Parent]"
test1.py:110: note: Only class variables allowed for class object access on protocols, parent is an instance variable of "Item"
test1.py:111: error: Argument 1 to "get_parent_name_attr2" has incompatible type "Item"; expected "Attr2Protocol"  [arg-type]
test1.py:111: note: Following member(s) of "Item" have conflicts:
test1.py:111: note:     parent: expected "ParentProtocol", got "Mapped[Parent]"
```

so what would be correct way of typing that protocol?

**Reply by CaselIT** [reply]
regarding question 1, I'm not too convinced that pyright behaviour is correct regarding a the "regular class variable" with no class value assigned.

Main issue with it is that it breaks every dataclass
```py
import dataclasses as DC

class WProtocol(Protocol):
    x: int

class W:
    x: int

    def __init__(self) -> None:
        self.x = 42

@DC.dataclass
class WDC:
    x: int

def takes_w(w: WProtocol) -> int:
    return w.x

takes_w(W)
takes_w(W())
takes_w(WDC)
takes_w(WDC(99))
```

this is accepted by pyright but raises an AttributeError at runtime.

**Reply by CaselIT** [reply]
I never noticed that pyright also allows
```py
WDC.x + 1
```

Is this a behaviour you are open to change?

**Reply by erictraut** [reply]
The authors of PEP 544 included `ClassVar` as a way to specify that a class-scoped variable should be treated as a "pure class variable". That is, it should be accepted as only a class variable, not as a instance variable, and it shouldn't be writable via a class instance because that would overwrite the class variable with an instance variable.

PEP 544 unfortunately provided no way to differentiate between a "pure instance variable" and "a class variable that can be shadowed by an instance variable". The PEP simply says "To distinguish between protocol class variables and protocol instance variables, the special `ClassVar` annotation should be used...". That implies that if a `ClassVar` annotation _isn't_ used, the variable should be considered an _instance variable_. I previously interpreted this to mean a "pure instance variable", but I later changed this interpretation. It was pointed out by pyright users that this interpretation is inconsistent with the way pyright treats non-protocol classes, and it's inconsistent with mypy's interpretation. I therefore changed pyright to treat a non-`ClassVar` variable within a protocol as a "class variable that can be shadowed by an instance variable".

Neither pyright nor mypy enforce that attributes (either class variables or instance variables) are assigned a value. This would unfortunately result in many false positives within typical Python code. Consequently, it is possible to write code that generates no type checking errors but will crash at runtime. That's the tradeoff.

Consider this code, which type checks without error in both mypy and pyright. Yet, if you call any of these four functions, the program will crash.

```python
from typing import Protocol

class WProtocol(Protocol):
    x: int

class W:
    x: int

def func1(w: WProtocol) -> int:
    return w.x

def func2(w: type[WProtocol]) -> int:
    return w.x

def func3(w: W) -> int:
    return w.x

def func4(w: type[W]) -> int:
    return w.x
```

**Reply by CaselIT** [reply]
Thanks you very much for the explanation.
Would having an hybrid approach work here by tracking what has a default value attached? Something like
```py
class W:
    x: int
    y: int = 42
    z: ClassVar[int] = 99

W.x # error, instance variable only
W.y # ok, regular classvar, that can be overridden on the instance
W.x # ok, pure classvar, can't be overridden on the instance
```

The above behavior it similar to how python works, and it's also compatible with how dataclasses work.

I'm not sure how to express that in a protocol. It seems you can assign a variable to a protocol, so maybe that could be a way, but I'm guessing that it's currently undefined how that should work for type checkers (both mypy and pyright seem to be fine with it)
It's probably fine to have the current behavior with Protocols though.

Re PEP 544, while I can't say I'm not familiar with it, it's really frustrating using the python typing since it's filled with feature that seem half backed to a mere user half backed, that work in one case but not in other very common ones.. I guessing that perfect is the enemy of good...

--- 
Returning to the original PropProtocol do you have a suggestion on how to express it correctly? See my additional unsuccessful tries here too https://github.com/microsoft/pyright/discussions/5915#discussioncomment-6952534

Again thank you for the time you dedicate to answering people questions

**Reply by erictraut** [reply]
> Would having an hybrid approach work here by tracking what has a default value attached?

The problem is that the following is perfectly legit Python code which runs fine and has no type violations. The hybrid approach that you suggest would introduce a false positive error.

```python
class W:
    x: int

W.x = 3
print(W.x)
```

> ...it's really frustrating using the python typing since it's filled with feature that seem half baked...

Some of this is a problem with type system features, but some of it is attributable to limitations of Python itself. Most programming languages don't allow a symbol to be defined without assigning it a value. Python doesn't enforce this in any way, and most Python code bases rely on this "loose" behavior. Most modern programming languages also have a keyword that allows a developer to indicate a symbol is `readonly`. With Python, there is no such keyword, although `Final` (introduced in [PEP 591](PEP 591)) partially addresses this shortcoming. Defining a read-only field in a protocol requires that you use a property, which is an imperfect stand-in for a read-only field, as you can see in your example above.

> Returning to the original PropProtocol do you have a suggestion on how to express it correctly?

I don't fully understand the problem you're trying to solve, so it's difficult to offer a suggestion. How is `parent` initialized in the `Parent` class? Is it intended to be a class variable or an instance variable or both? Is it assigned only once, or is it modified in multiple places? Is `Parent` an abstract class and there are concrete subclasses that you're not showing in your example? Would a generic class be a more appropriate way to model this problem?

I assume that these descriptors are all meant to be assigned to class variables because member accesses through an instance variable will not invoke the descriptor protocol. I also assume that the descriptors are assigned once at class definition time rather than dynamically after the class is defined. If my assumptions are correct, then all of your attributes should be marked explicitly as `Final` both within your protocol definitions and non-protocol class definitions.

```python
class ParentProtocol(Protocol):
    name: Final[Mapped[str]] = Mapped()

class Parent:
    name: Final[Mapped[str]] = Mapped()

class Item:
    id: Final[Mapped[int]] = Mapped()
    parent: Final[Mapped[Parent]] = Mapped()

class ItemProtocol(Protocol):
    parent: Final[Mapped[ParentProtocol]] = Mapped()

def get_parent_name(child: ItemProtocol) -> str:
    return child.parent.name

get_parent_name(Item())
```

**Reply by CaselIT** [reply]
> > Would having an hybrid approach work here by tracking what has a default value attached?
> 
> The problem is that the following is perfectly legit Python code which runs fine and has no type violations. The hybrid approach that you suggest would introduce a false positive error.
> 
> ```python
> class W:
>     x: int
> 
> W.x = 3
> print(W.x)
> ```

Sure, the hybrid approach does a different tradeoff.
One could argue that it's a better one, since I can't think of a common reason to set a classvar after the class definition, other than in some special circumstances such as to avoid a circular import. I believe that declaring instance variables in the class is instead a very common pattern given the popularity of dataclasses and the fact that it's a nice way of giving them documentation that is picked up by sphinx and other doc generation tools and also ides.

> > ...it's really frustrating using the python typing since it's filled with feature that seem half baked...
> 
> Some of this is a problem with type system features, but some of it is attributable to limitations of Python itself. Most programming languages don't allow a symbol to be defined without assigning it a value. Python doesn't enforce this in any way, and most Python code bases rely on this "loose" behavior. Most modern programming languages also have a keyword that allows a developer to indicate a symbol is `readonly`. With Python, there is no such keyword, although `Final` (introduced in [PEP 591](PEP 591)) partially addresses this shortcoming. Defining a read-only field in a protocol requires that you use a property, which is an imperfect stand-in for a read-only field, as you can see in your example above.

Indeed, sorry, that comment was a bit unwarranted, since I imagine that it's all but easy to properly came up with the constructs that statically model the types within python.
> 
> > Returning to the original PropProtocol do you have a suggestion on how to express it correctly?
> 
> I don't fully understand the problem you're trying to solve, so it's difficult to offer a suggestion. How is `parent` initialized in the `Parent` class? Is it intended to be a class variable or an instance variable or both? Is it assigned only once, or is it modified in multiple places? Is `Parent` an abstract class and there are concrete subclasses that you're not showing in your example? Would a generic class be a more appropriate way to model this problem?
> 
> I assume that these descriptors are all meant to be assigned to class variables because member accesses through an instance variable will not invoke the descriptor protocol. I also assume that the descriptors are assigned once at class definition time rather than dynamically after the class is defined. If my assumptions are correct, then all of your attributes should be marked explicitly as `Final` both within your protocol definitions and non-protocol class definitions.
> 
> ```python
> class ParentProtocol(Protocol):
>     name: Final[Mapped[str]] = Mapped()
> 
> class Parent:
>     name: Final[Mapped[str]] = Mapped()
> 
> class Item:
>     id: Final[Mapped[int]] = Mapped()
>     parent: Final[Mapped[Parent]] = Mapped()
> 
> class ItemProtocol(Protocol):
>     parent: Final[Mapped[ParentProtocol]] = Mapped()
> 
> def get_parent_name(child: ItemProtocol) -> str:
>     return child.parent.name
> 
> get_parent_name(Item())
> ```

In general the `Mapped` descriptor is declared in two way: it's either manually assigned when some additional sql setting are needed, like indicate that a field has a primary key, or only the left side can be declared. In both cases sqlalchemy will ensure that the proper descriptor is created when the class is defined using `__init_subclass__`. 

In general it's only assigned once at the class level. An instance can instead assign a value in the init or later. In the example Parent is a real class. There may be cases where there are also subclasses of Parent sure, but it's less common. 

Here is an example taken from https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#declarative-mapping
```py
from sqlalchemy import Integer, String, ForeignKey
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


# declarative base class
class Base(DeclarativeBase):
    pass


# an example mapping using the base
class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    fullname: Mapped[str] = mapped_column(String(30))
    nickname: Mapped[Optional[str]]
```

here the descriptor of `name` and `nickname` is assigned when `User` is defined by the `DeclarativeBase.__init_subclass__` method.

I've tried your proposal with final, that would require that name and nickname are given assinged, and everything seems to be working correctly, meaning that I can use the properties from the class to access the `InstrumentedAttribute` to build sql expressions, and at the instance level I can assign values etc correctly.

I think we can probably document this as a way of better defining the types of the model. Thanks.
I've a couple of question:
Can `Mapped` by made `Final` automatically, without need to add `Final[Mapped[...]]`?
Mypy does not seem to like final on protocols (`error: Protocol member cannot be final  [misc]`) but I'm guessing it's a mypy issue.


Thanks again for the reply

**Reply by erictraut** [reply]
> Can `Mapped` by made `Final` automatically, without need to add `Final[Mapped[...]]`?

I'm not sure what you mean. `Final` doesn't apply to the _type_ of the attribute (which in this case is `Mapped[...]`). The `Final` annotation is meta-information about the _attribute_ independent of its type. So no, `Final` can't be applied automatically. You need to specify `Final`, just like you would need to provide a `readonly` modifier if the language supported such a keyword.

> Mypy does not seem to like final on protocols

Yeah, that's a mypy issue. You'll need to take it up with the mypy maintainers. I can't think of any justification for this limitation, but maybe they are thinking of some reason that's eluding me.

I noted that pyright currently requires you to assign a value to a `Final` variable within a protocol definition. That seems unnecessary, so I just changed the behavior to eliminate this check within protocols. That means you can remove the unnecessary `= Mapping()` within the two protocol class definitions in my above example.

**Reply by CaselIT** [reply]
> > Can `Mapped` by made `Final` automatically, without need to add `Final[Mapped[...]]`?
> 
> I'm not sure what you mean. `Final` doesn't apply to the _type_ of the attribute (which in this case is `Mapped[...]`). The `Final` annotation is meta-information about the _attribute_ independent of its type. So no, `Final` can't be applied automatically. You need to specify `Final`, just like you would need to provide a `readonly` modifier if the language supported such a keyword.

Ok, I was basically asking if something like this can be done (currently `Final` is just ignored)
```py
_T = TypeVar("_T", covariant=True)
MappedFinal = Final[Mapped[_T]]
```

The current behavior is probably the most sensible one

> > Mypy does not seem to like final on protocols
> 
> Yeah, that's a mypy issue. You'll need to take it up with the mypy maintainers. I can't think of any justification for this limitation, but maybe they are thinking of some reason that's eluding me.
> 
> I noted that pyright currently requires you to assign a value to a `Final` variable within a protocol definition. That seems unnecessary, so I just changed the behavior to eliminate this check within protocols. That means you can remove the unnecessary `= Mapping()` within the two protocol class definitions in my above example.

makes sense, thanks. I'll try experimenting a bit more on this and look if the approach of using final can be documented within sqlalchemy. Thanks for the suggesting a solution


Regarding the "hybrid approach" mentioned above, would a new discussion or issue be a better place to further discuss it, or you think it would not be worth doing?

**Reply by erictraut** [reply]
What you're calling a "hybrid approach" would result in too many false positive errors, so I think it's unlikely that pyright or mypy would implement it, at least in the manner you've suggested. As I mentioned, it's common in Python to assign a value to a class or instance variable later, outside of a class' definition. There is nothing wrong with this from a type perspective.

A while ago, I added a check to pyright called [reportUninitializedInstanceVariable](https://microsoft.github.io/pyright/#/configuration?id=diagnostic-rule-defaults). It enforces that all instance variables declared in a class body are assigned a value either within the class body or the constructor. Perhaps that's the check you're looking for. To my knowledge, very few pyright users enable this feature because it forces a coding style that is not typical for most Python code bases. But if you're willing to adopt this coding pattern throughout your code, you will avoid the class of bug that you're concerned about here.

```python
# pyright: reportUninitializedInstanceVariable=true

class W1:
    x: int # Error: x is not initialized

class W2:
    x: int = 2 # OK

class W3:
    def __init__(self):
        self.x = 3 # OK
```

**Reply by CaselIT** [reply]
Fair enough. I'm not sure I agree with "it's common in Python to assign a value to a class outside of a class' definition" but I guess the preference is to prefer false nagatives compared to false positive

I don't think that flag is useful to me, since it does not allow dataclasses (unless all variables have a default). Thanks anyway

**Reply by erictraut** [reply]
> I don't think that flag is useful to me, since it does not allow dataclasses

It does support dataclasses. The synthesized dataclass `__init__` method assigns values to the instance variables that are missing a default, so these are considered initialized.

**Reply by CaselIT** [reply]
ok, understood
